### PR TITLE
638 Feature Request: password generator with options

### PR DESCRIPTION
**Pull request by Christian-Oleson**
## Pull request type
Feature Request

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] UI change (please include screenshot!)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Internationalization and localization
- [ ] Other (please describe):

## What is the current behavior?
N/a

Issue Number: 638

## What is the new behavior?
A new password generator feature has been made. You can select
- Password Length
- To include lowercase characters
- To include uppercase characters
- To include special characters
- How many passwords to generate

## Other information
- My first PR to this repo
- I am looking for initial feedback but may want to add Spanish and German translations to start, though I am unsure if you need all translations to be added up front.

## Quality check

Before creating this PR, have you:

- [x] Followed the code style guideline as described in [CONTRIBUTING.md](https://github.com/veler/DevToys/blob/main/CONTRIBUTING.md)
- [x] Verified that the change work in Release build configuration
- [ ] Checked all unit tests pass

### CODE REVIEW

File: src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    <value>Passwords</value>
+  </data>
+  <data name="MultiplySymbol" xml:space="preserve">
+    <value>x</value>
+  </data>
+  <data name="NumberOfPasswordsToGenerate_AutomationProperties_Name" xml:space="preserve">
+    <value>Number of Passwords to generate</value>
+  </data>
+  <data name="Uppercase" xml:space="preserve">
+    <value>Include Lowercase Characters in a Password (ABCDEFGHJKLMNOPQRSTUVWXYZ)</value>
```
Comment (veler) [inline src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw:145]: `Lowercase` should be `Uppercase`.
Also, I would recommend displaying the `(ABCDEFGHJKLMNOPQRSTUVWXYZ)` in the `Description` property of the `ExpandableSettingControl`, which will display the smaller text like here
image

Comment (veler) [reply src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw:145]: Same thing for Digits, Lowercase and SpecialCharacters

File: src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  </data>
+  <data name="Description" xml:space="preserve">
+    <value>Generate random passwords</value>
+  </data>
+  <data name="Digits" xml:space="preserve">
```
Comment (veler) [inline src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw:150]: I don't see where the Digits text is displayed. More generally, it looks like there's no "Digits" option in the UI. Should this be removed?

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Composition;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs:1]: Please add `#nullable enable` at the top of the file.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    [ProtocolName("password")]
+    [Order(1)]
+    [CompactOverlaySize(width: 400, height: 500)]
+    internal class PasswordGeneratorToolProvider : IToolProvider
+    {
+        private readonly IMefProvider _mefProvider;
+
+        public string MenuDisplayName => LanguageManager.Instance.PasswordGenerator.MenuDisplayName;
+
+        public string? SearchDisplayName => LanguageManager.Instance.PasswordGenerator.SearchDisplayName;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs:19]: This string (and the other below) shouldn't be nullable.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,271 @@
+﻿using System;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:1]: nullable enable :)

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        private static readonly SettingDefinition<bool> Lowercase
+            = new(
+                name: $"{nameof(PasswordGeneratorToolViewModel)}.{nameof(Lowercase)}",
+                isRoaming: true,
+                defaultValue: true);
+
+        /// <summary>
+        /// Whether the password should include numbers.
+        /// </summary>
+        private static readonly SettingDefinition<bool> Numbers
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:42]: Is it really used?

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        {
+            get => _output;
+            set => SetProperty(ref _output, value);
+        }
+
+        [ImportingConstructor]
+        public PasswordGeneratorToolViewModel(ISettingsProvider settingsProvider, IMarketingService marketingService)
+        {
+            SettingsProvider = settingsProvider;
+            _marketingService = marketingService;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:185]: I think we can remove this service for this tool. There's no much utility for it.

File: src/dev/impl/DevToys/Views/Tools/Generators/PasswordGenerator/PasswordGeneratorPage.xaml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    <Grid Margin="0,0,0,16" RowSpacing="12">
+        <VisualStateManager.VisualStateGroups>
+            <VisualStateGroup>
+                <VisualState>
+                    <VisualState.StateTriggers>
+                        <ex:IsCompactOverlayModeTrigger TargetElement="{x:Bind}" />
+                    </VisualState.StateTriggers>
+                    <VisualState.Setters>
+                        <Setter Target="LengthSetting.Visibility" Value="Collapsed" />
```
Comment (veler) [inline src/dev/impl/DevToys/Views/Tools/Generators/PasswordGenerator/PasswordGeneratorPage.xaml:21]: This is great, but it doesn't hide the `Configuration` text.
I'd suggest to give a name to the StackPanel on line 42 and collapse the whole StackPanel  on line 21 (and remove lines 22 to 24).

File: src/dev/shared/SharedAssemblyInfo.cs [hunk 9]
```diff
@@ -37,7 +37,7 @@
 
 // Please DO NOT commit changes on the 2 lines below unless
 // you're about to release a new version of the app on the Store.
-[assembly: AssemblyVersion("1.0.9.0")]
-[assembly: AssemblyFileVersion("1.0.9.0")]
+[assembly: AssemblyVersion("1.0.10.0")]
```
Comment (veler) [inline src/dev/shared/SharedAssemblyInfo.cs:40]: Please revert the changes in this file :)

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs (hunk 6 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:42]: I think I originally meant to add the numbers, but then realize the feature request didn't have anything about whether or not to include an option of include/exclude numbers. Would we like to do so? Or should we omit them  from the options?

File: src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw (hunk 2 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw:150]: Please see this comment:
https://github.com/veler/DevToys/pull/664/files#r975456968

File: src/dev/DevToys.Startup/Package.appxmanifest [hunk 10]
```diff
@@ -15,7 +15,7 @@
     <Identity
         Name="DevToysPreview"
         Publisher="CN=etiennebaudoux"
-        Version="1.0.9.0" />
+        Version="1.0.10.0" />
```
[the remark this answers is not in the document]
Comment (Christian-Oleson) [reply src/dev/DevToys.Startup/Package.appxmanifest:18]: reverted

File: src/dev/impl/DevToys/Helpers/CryptoRandomGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+using System.Security.Cryptography;
+
+namespace DevToys.Helpers
+{
+    /// <summary>
+    /// A class that mimics the standard Random class in the .NET Framework - but uses a random number generator internally.
+    /// Taken from IdentityModel (ref.: https://github.com/IdentityModel/IdentityModel/ )
+    /// Taken from PasswordGenerator (ref.: https://github.com/Darkseal/PasswordGenerator/blob/master/CryptoRandom.cs )
+    /// </summary>
+    public class CryptoRandom : Random
```
[the remark this answers is not in the document]
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/Helpers/CryptoRandomGenerator.cs]: done

File: src/dev/impl/DevToys/Helpers/CryptoRandomGenerator.cs [hunk 12]
```diff
@@ -0,0 +1,228 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/Helpers/CryptoRandomGenerator.cs]: done

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs (hunk 3 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs]: done

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs (hunk 4 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolProvider.cs]: fixed

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs (hunk 7 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs]: done

File: src/dev/shared/SharedAssemblyInfo.cs (hunk 9 above)
Comment (Christian-Oleson) [reply src/dev/shared/SharedAssemblyInfo.cs]: done

File: src/dev/impl/DevToys/DevToys.csproj [hunk 13]
```diff
@@ -576,6 +582,7 @@
     <AppxManifest Include="Package.appxmanifest">
       <SubType>Designer</SubType>
     </AppxManifest>
+    <PRIResource Include="Strings\en-US\PasswordGenerator.resw" />
```
[the remark this answers is not in the document]
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/DevToys.csproj:587]: done

File: src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw (hunk 1 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/Strings/en-US/PasswordGenerator.resw]: done

**Comment by Christian-Oleson** [comment]
> Hello, Thank you for this PR :D It looks great overall.
> 
> To give you some initial feedback and answer your questions:
> 
> 1. No need to translate the tool yet. We use Crowdin (https://crowdin.com/project/devtoys), which will automatically pull the added strings from `en-us` in the `main` branch. That means we should first approve and merge this PR before doing the translation.
> 2. Since this is a new tool, we'd need to create a new icon for it. If you're willing to do it, you can try to follow the few instructions I left here: https://github.com/veler/DevToys/blob/main/assets/font/README.md
>    Initially, these instructions were for me, as I'm not a pro a making icons and fonts. If you don't feel like doing the icon, let me know and I will happily do it.
>    An idea of icon: how about something with `***` or
>    image
>    This icon comes from [FluentSystemIcons-Regular](https://github.com/microsoft/fluentui-system-icons) (glyph `F59E`) which we could copy and/or edit since it's under MIT license.

I'll have to see if I can find another computer. FontForge is not liked by my machine it seems. I'll work to get it on another and see if I can follow the setup to generate a

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs (hunk 6 above)
Comment (veler) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:43]: So, currently we have the possibility to generate password _**without**_:
- lowercase letters
- uppercase letters
- symbols
If we also remove numbers... well there's not much to generate then 😅

But I do see the value for generating a password without numbers too.
In term of UX, I'm quite undecided at the moment, but in short, assuming we have 4 options
- lowercase letters
- uppercase letters
- symbols
- numbers
we should enforce at least 1 option, or display a message asking the user to enable at least 1 option in order to generate a password.

These options could be displayed as:
- Toggle check box (like today)
- List with multiple selection
- List of check box
- else? (I can't think of anything else)

What do you think?

Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:43]: It would be very easy to display a message in the output text if no options are selected. The other options would require more work, but also would be acceptable.

Comment (veler) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:43]: Could we give a try at the displaying a message?
Perhaps you can use an `InfoBar` like here: https://github.com/veler/DevToys/blob/2e99e04d6f363a5bddf044afc5545266be43246c/src/dev/impl/DevToys/Views/Tools/Text/RegEx/RegExToolPage.xaml#L118-L123 But display a `Warning` or `Information` instead of `Error`?

Another option I see (but it might be more complex too) is that when it remains only 1 option checked, we disable the checked option (the option will be checked but user won't be able to uncheck it) until another option is checked.

File: src/dev/impl/DevToys/Views/Tools/Generators/PasswordGenerator/PasswordGeneratorPage.xaml (hunk 8 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/Views/Tools/Generators/PasswordGenerator/PasswordGeneratorPage.xaml:21]: I believe I made the modification desired here.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs (hunk 6 above)
Comment (Christian-Oleson) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/PasswordGenerator/PasswordGeneratorToolViewModel.cs:43]: I'll take a look at this. It just may be awhile since I've gotten a bit busy.

**Comment by Christian-Oleson** [comment]
FYI, I'm still working this, just went out on parental leave last week, so it will take a bit longer than expected.

**Comment by veler** [comment]
> FYI, I'm still working this, just went out on parental leave last week, so it will take a bit longer than expected.

@Christian-Oleson no worries :) I can address the remaining feedback myself if you'd like too.
Also, congratulations for the newborn! 💯

**Comment by nj** [comment]
Might be good to extend further later on, perhaps with inspiration by https://randomkeygen.com/ (on GitHub at https://github.com/circlecell/randomkeygen.com, MIT licensed).

**Comment by Christian-Oleson** [comment]
> > FYI, I'm still working this, just went out on parental leave last week, so it will take a bit longer than expected.
> 
> 
> 
> @Christian-Oleson no worries :) I can address the remaining feedback myself if you'd like too.
> 
> Also, congratulations for the newborn! 💯 

Thank you! It may be best to have you finish. I'm still out and a bit busy, so it may be best to get this released instead of having folks wait on me eventually completing this on my end.

**Comment by veler** [comment]
Closing this PR as the tool got completed into #873 . Thank you @Christian-Oleson  and @micahmo  for this contribution! :)
